stdlib/Base64/src/decode.jl

```diff
@@ -32,14 +41,15 @@ julia> String(read(iob64_decode))
 "Hello!"
 ```
 """
-struct Base64DecodePipe <: IO
+struct Base64DecodePipe{F<:Function} <: IO
     io::IO
     buffer::Buffer
+    decode::F
     rest::Vector{UInt8}
 
-    function Base64DecodePipe(io::IO)
+    function Base64DecodePipe(io::IO; decode::T=decodeonechar_base64) where {T<:Function}
```

Comment:
I'm not thrilled with the API adding an undocumented keyword where you are supposed to pass an undocumented internal function.

Reply:
If we are going to add a parameter to the constructor, maybe just an (exported) singleton type like `Base64()` vs `Base64Url()` (which could be passed to either `Base64DecodePipe` or `Base64EncodePipe`).  Then, the `decode` and `encode` functions could take this parameter and dispatch on it.  And you could also pass it to `base64encode` etcetera rather than exporting a new function.

Reply:
Thank you for the review. I was not well considering a good API signature.

The problem I encountered when trying to give different types (`Base64()` and `Base64Url()` in above example) as an argument to the constructor is type instability in `Base64DecodePipe`, since I was removing type parameters for the reason pointed out in the above comment (https://github.com/JuliaLang/julia/pull/49849#discussion_r1202974440). But I think it is also true that `Base64DecodePipe` (and also `Base64EncodePipe`) needs to contain information on an encoding format (base64 or base64url).

So, for this purpose, is it acceptable to use an exported `enum` (like `@enum Base64Format BASE64 BASE64URL`) as an parameter for the constructor, in the way I tried in ebc21ed?

Reply:
An enum could be fine, although of course there is then a runtime branch.   If you make the enum value `BASE64 = 0` and `BASE64URL = 256`, however, then you could simply make your encoder/decoder lookup tables of length 512 and add the encoder type as offset in the lookup.   This would just add a single addition to the encoding cost, which is probably negligible.

Reply:
I see, then it seems the previous commit is still fine, for actually ebc21ed implements the way you suggested.
`@benchmark` results in small samples seems to show no performance regression so far.